querier.go

```diff
@@ -904,6 +911,14 @@ func (s *chainedSeries) Iterator() SeriesIterator {
 	return newChainedSeriesIterator(s.series...)
 }
 
+func (s *chainedSeries) ChunkIterator() ChunkIterator {
+	ch := &chainedChunkIterator{chain: make([]ChunkIterator, 0, len(s.series))}
+	for _, s := range s.series {
+		ch.chain = append(ch.chain, s.ChunkIterator())
```

Comment:
What are the ordering guarantees here? I'd expect this to all be in order, reencoding if needed.

Reply:
Good point, we need to follow on what `Iterator` has I guess.

Reply:
Ok I looked it through and to me `chainedSeriesIterator` or even `mergedSeriesSet` does not do any sorting on given series. IMO we should not do it here either in that case. I added comments that we expect sorted and non overlapping chunks.

Reply:
verticalChainedSeries does though.

Reply:
Good point, missed that. :man_facepalming:  

What if we will return error (not implemented) for now for `verticalChainedSeries`? It is used by vertical compactor and we can add this implementation only when we will be starting using `ChunkIterations` for that (if any).

My point is to iterate with this faster, so we can gain benefits in remote read streaming (:

Reply:
I thought that using that was automatic if overlapping blocks were present? If so that'd break the remote read endpoint.

Reply:
It is in indeed: https://github.com/prometheus/tsdb/blob/d5b3f0704379a9eaca33b711aa0097f001817fc2/db.go#L1185

Ok, let's not be lazy and try to do overlaping chunk iterators properly then :thinking:

Reply:
I can see current policy for vertical merging is merging both series and fill the samples from both sides. This breaks all the intervals we used to have if the timestamps are slightly off for series a vs b. Is it intentional? If the timestamps are the same, we just choose arbitrary ones (b).

https://github.com/prometheus/tsdb/blob/d097d3fbf4c87463d3300f897182b269eefd5c30/querier.go#L1061 

In the same how would you treat overlapping chunks in such vertical case?

I think the outcome would be to have non-overlapping flow of chunks for the `ChunkIterator` so we need to reencode two chunks into max 120 samples new chunks probably reusing `verticalMergeSeriesIterator`... let me know if my thinking is wrong.

@codesome @brian-brazil

Reply:
Side question:

We habe in `chunks`: 
```
// Iterator is a simple iterator that can only get the next value.
type Iterator interface {
	At() (int64, float64)
	Err() error
	Next() bool
}

```
vs
```
// SeriesIterator iterates over the data of a time series.
type SeriesIterator interface {
	// Seek advances the iterator forward to the given timestamp.
	// If there's no value exactly at t, it advances to the first value
	// after t.
	Seek(t int64) bool
	// At returns the current timestamp/value pair.
	At() (t int64, v float64)
	// Next advances the iterator by one.
	Next() bool
	// Err returns the current error.
	Err() error
}
```

Should I aim for having chunk iterators implement `Seek` or just convert chunk iterator to the one with `nopSeek` (returning err not implemented). ?

Reply:
It wouldn't hurt to have a very simple seek implementation, it doesn't need to be efficient.

Reply:
fair

Reply:
Seek implementation added.